x/session/keeper/session_hydrator.go

```diff
@@ -0,0 +1,203 @@
+	sharedtypes "pocket/x/shared/types"
+)
+
+var SHA3HashLen = crypto.SHA3_256.Size()
+
+// TODO(#21): Make these configurable governance param
+const (
+	NumBlocksPerSession   = 4
+	NumSupplierPerSession = 15
+)
+
+type sessionHydrator struct {
+	// The session header that is used to hydrate the reset of the session data
+	sessionHeader *types.SessionHeader
+
+	// The fully hydrated session object
+	session *types.Session
+
+	// The height at which the session being request
+	blockHeight int64
+
+	// A redundant helper that maintains a hex decoded copy of `session.Id` used for session hydration
+	sessionIdBz []byte
+}
+
+func NewSessionHydrator(
+	appAddress string,
+	serviceId string,
+	blockHeight int64,
+) *sessionHydrator {
+	sessionHeader := &types.SessionHeader{
+		ApplicationAddress: appAddress,
+		ServiceId:          &sharedtypes.ServiceId{Id: serviceId},
+	}
+	return &sessionHydrator{
+		sessionHeader: sessionHeader,
+		session:       &types.Session{},
+		blockHeight:   blockHeight,
+		sessionIdBz:   make([]byte, 0),
+	}
+}
+
+// GetSession implements of the exposed `UtilityModule.GetSession` function
+// TECHDEBT(#519): Add custom error types depending on the type of issue that occurred and assert on them in the unit tests.
+func (k Keeper) HydrateSession(ctx sdk.Context, sh *sessionHydrator) (*types.Session, error) {
+	logger := k.Logger(ctx).With("method", "hydrateSession")
+
+	if err := k.hydrateSessionMetadata(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate the session metadata: %v", err)
+	}
+	logger.Debug("Finished hydrating session metadata")
+
+	if err := k.hydrateSessionID(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate the session ID: %v", err)
+	}
+	logger.Info("Finished hydrating session ID: %s", sh.sessionHeader.SessionId)
+
+	if err := k.hydrateSessionApplication(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate application for session: %v", err)
+	}
+	logger.Debug("Finished hydrating session application: %+v", sh.session.Application)
+
+	if err := k.hydrateSessionSuppliers(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate suppliers for session: %v", err)
+	}
+	logger.Debug("Finished hydrating session suppliers: %+v")
+
+	sh.session.Header = sh.sessionHeader
+	sh.session.SessionId = sh.sessionHeader.SessionId
+
+	return sh.session, nil
+}
+
+// hydrateSessionMetadata hydrates metadata related to the session such as the height at which the session started, its number, the number of blocks per session, etc..
+func (k Keeper) hydrateSessionMetadata(ctx sdk.Context, sh *sessionHydrator) error {
+	sh.session.NumBlocksPerSession = NumBlocksPerSession
+	sh.session.SessionNumber = int64(sh.blockHeight / NumBlocksPerSession)
+	sh.sessionHeader.SessionStartBlockHeight = sh.blockHeight - (sh.blockHeight % NumBlocksPerSession)
```

Comment:
Assuming Session#1 starts at BlockHeight#1 (and there is no Session#0 nor Block#0), I think we should have a `+1` here

```suggestion
	sh.sessionHeader.SessionStartBlockHeight = sh.blockHeight - (sh.blockHeight % NumBlocksPerSession) + 1
```

Reply:
This should not have a +1

Reply:
If the first block ever is `0` and the first session ever is `0` then all good. But it seems that the first session ever is `1` which makes the mis-alignment weird

Reply:
> But it seems that the first session ever is 1 which makes the mis-alignment weird

That's what I had originally, so I'm going to change it back.